capi-matrix/README.md

```diff
@@ -0,0 +1,50 @@
+# RFC 0015 - CAPI matrix per provider
+
+## Introduction
+
+We have a cluster-<provider> chart for every provider that we support. It has a set of templates that allow our customers to create a cluster app with all necessary resources required by a Cluster API implementation. 
+
+In the values of these charts we define the Kubernetes version of the cluster and the image that machine will use to run. These images contain the Kubernetes components with a specific version. 
+
+One of the goal in this RFC is decouple Kubernetes version from image names so customer does not need to know which image correspond to which version. At the same time we might want to provide new OS version when new CVE appears without necessarily change the Kubernetes version. Finally having a list of kubernetes version certified as ready to be used benefit to disallow users to create cluster in flawy versions.
+
+## User stories
+
+- As a customer, I want to not worry about the kubernetes version and rely on Giant Swarm default version provided.
+
+- As Giant Swarm staff, I don't want to expend time upgrading cluster manually.
+
+- As Giant Swarm staff, I would like to forbid a specific Kubernetes/OS version to be used in case of security constraints.
+
+- As customer, I want to have the option to have a cluster pinned to a version.
+
+## Solution
+
+- Lets have a cluster-<provider> chart release for every Kubernetes version is released. That way our automation can roll upgrades automatically when new version is published
```

Comment:
I think this is the wrong approach.

I _still_ think we need an abstraction on top of cluster apps that the user deals with. That layer should include this kind of pinning.

Reply:
We proposed a meta app in the past and I think we discussed it in Kass sync. There were some drawbacks and it was discarded for now. What is the problem of using a kuberneteVersion default in this layer? BTW CAPG and CAPA is doing too

Reply:
but this is not limiting or changing anything about your approach. It just makes easier the customer life, defaulting the image name based on the kubernetes version (at the same time allow us to control which versions are available to the customer)

Reply:
What Marcus is saying is that there should be just one version available. So neither the customer should be able to pin any version nor we provide any choice. I'd like to get there too. 

But how would we solve freeze periods then? Do we have confidence that we won't break anything and everything is just kept up to date whatsoever? Or do we still need to be able to stay at a certain version? And will we be able to test the upgrade that happens after the freeze period? This might be a bigger jump.

Reply:
IMO forcing a version per cluster_PROVIDER release limits us not the customer. Let me explain it. In our whole history we have been managing clusters and apps, and in many times across all customers we have some requirements, blockers,... that block us to upgrade to certain versions of an app or a k8s (even OS when cgroups).

This is going to happen no matter how good the cluster upgrade mechanics are. You also mention other conditions like frozen periods or software dependencies (openet ehem ehem) that force us to offer that option.

IMO the problem we had in the past is the manual toil of schedule, discuss, manual preparation, ... which required us to do an upgrade. This is what we need to solve and for that reason [the automate cluster upgrade rfc](https://github.com/giantswarm/rfc/tree/main/automatic-cluster-upgrades) was created. 

I don't say that we need to put the list of supported versions in cluster-PROVIDER charts, we could find a better place, but we will need that list somewhere. Leaving the chart without the default k8s/OS version and defining it in another place for all providers is another option. But we need to clarify the vision of how clusters/app should be updated.